File: base_station/src/base_station/web/static/daq-config/validation.js
import { differentialNegative } from './channels.js';
import { daqConnectionAllowed } from './connection-policy.js';
import { isSpecNodeType, validateSpecNode } from './node-specs.js';

const MEASUREMENT_TYPES = new Set(['labjack-ain', 'labjack-current', 'labjack-thermocouple']);
const HARDWARE_NODE_TYPES = new Set([
  'labjack-channel', 'labjack-channel-pair', ...MEASUREMENT_TYPES,
  'pressure-calibration', 'load-cell',
]);

/**
 * Fast client-side checks. The server repeats authoritative validation on save.
 * @param {Object} graph
 * @param {Object} labjackSettings
 * @returns {{severity:'error'|'warning',subject:string,message:string}[]}
 */
export function validateGraph(graph, labjackSettings = {}) {
  const issues = [];
  const nodes = graph?.nodes ?? [];
  const incoming = new Map();
  for (const link of graph?.links ?? []) {
    incoming.set(`${link.toNode}:${link.toPin}`, link);
  }
  validateLinks(graph, issues);
  const mux = Boolean(labjackSettings?.mux80Enabled);
  const usedSources = new Map();

  for (const node of nodes) {
    if (!isSpecNodeType(node.nodeType) && !HARDWARE_NODE_TYPES.has(node.nodeType)) {
      issues.push(issue('error', node.id, `Unsupported node type: ${node.nodeType}`));
    }
    if (node.nodeType === 'labjack-channel' || node.nodeType === 'labjack-channel-pair') {
      validateChannel(node, mux, usedSources, issues);
    }
    if (MEASUREMENT_TYPES.has(node.nodeType)) validateMeasurement(node, graph, incoming, issues);
    for (const pin of node.pins ?? []) {
      if (pin.direction !== 'input' || pin.optional) continue;
      if (!incoming.has(`${node.id}:${pin.id}`)) {
        issues.push(issue('warning', node.id, `${pin.label} is not connected`));
      }
    }
    const specMessages = validateSpecNode(node);
    if (specMessages) {
      for (const message of specMessages) issues.push(issue('error', node.id, message));
    }
    if (node.nodeType === 'pressure-calibration') {
      let complete = true;
      for (const field of ['inputMin', 'inputMax', 'psiMin', 'psiMax']) {
        if (!requiredNumber(linkedOrLiteral(node, field, graph, incoming, node.config?.[field]))) {
          issues.push(issue('error', node.id, 'Pressure calibration requires all four calibration values'));
          complete = false;
          break;
        }
      }
      const inputMin = linkedOrLiteral(node, 'inputMin', graph, incoming, node.config?.inputMin);
      const inputMax = linkedOrLiteral(node, 'inputMax', graph, incoming, node.config?.inputMax);
      if (complete && Number(inputMin) === Number(inputMax)) {
        issues.push(issue('error', node.id, 'Pressure calibration input span cannot be zero'));
      }
    }
    if (node.nodeType === 'load-cell') {
      const rated = linkedOrLiteral(node, 'ratedOutputMvV', graph, incoming, node.config?.ratedOutputMvV);
      const capacity = linkedOrLiteral(node, 'capacity', graph, incoming, node.config?.capacity);
      const zero = linkedOrLiteral(node, 'zeroV', graph, incoming, node.config?.zeroV);
      const excitation = linkedOrLiteral(node, 'excitation', graph, incoming, node.config?.excitationV);
      if (!(Number(rated) > 0)) issues.push(issue('error', node.id, 'Rated output must be positive'));
      if (!(Number(capacity) > 0)) issues.push(issue('error', node.id, 'Rated capacity must be positive'));
      if (!requiredNumber(zero)) {
        issues.push(issue('error', node.id, 'Load cell zero offset is required'));
      }
      if (!(Number(excitation) > 0)) issues.push(issue('error', node.id, 'Load cell excitation must be positive'));
    }
  }
  return issues;
}

function validateLinks(graph, issues) {
  const nodes = new Map((graph?.nodes ?? []).map((node) => [node.id, node]));
  for (const link of graph?.links ?? []) {
    const sourceNode = nodes.get(link.fromNode);
    const targetNode = nodes.get(link.toNode);
    const source = sourceNode?.pins?.find((pin) => pin.id === link.fromPin);
    const target = targetNode?.pins?.find((pin) => pin.id === link.toPin);
    if (!sourceNode || !targetNode || !source || !target) continue;
    if (daqConnectionAllowed(source, target, sourceNode, targetNode, graph)) continue;
    issues.push(issue(
      'error',
      targetNode.id,
      `${target.label} cannot accept ${source.type ?? 'this connection'}`,
    ));
  }
}

function validateChannel(node, mux, usedSources, issues) {
  const channel = node.config?.channel;
  const number = /^AIN\d+$/.test(channel ?? '') ? Number(channel.slice(3)) : NaN;
  const valid = Number.isInteger(number) && (number <= 13 || (mux && number >= 48 && number <= 127));
  if (!valid || (mux && number >= 4 && number <= 13)) {
    issues.push(issue('error', node.id, `${channel ?? 'Channel'} is unavailable for this hardware setup`));
  }
  if (node.nodeType === 'labjack-channel-pair') {
    const negative = differentialNegative(channel);
    const negativeNumber = /^AIN\d+$/.test(negative ?? '') ? Number(negative.slice(3)) : NaN;
    const validPositive = Number.isInteger(number) && (
      (number < 16 && number % 2 === 0)
      || (number >= 48 && number <= 127 && Math.floor((number - 48) / 8) % 2 === 0)
    );
    if (!validPositive || !Number.isInteger(negativeNumber)) {
      issues.push(issue('error', node.id, `${channel ?? 'Channel'} cannot start a differential pair`));
    }
  }
  const key = channel;
  const previous = usedSources.get(key);
  if (previous && previous !== node.id) {
    issues.push(issue('warning', node.id, `${channel} is also referenced by ${previous}`));
  } else usedSources.set(key, node.id);
}

function validateMeasurement(node, graph, incoming, issues) {
  const config = node.config ?? {};
  if (node.nodeType === 'labjack-thermocouple' && !node.config?.thermocoupleType) {
    issues.push(issue('error', node.id, 'Select the thermocouple type'));
  }
  if (node.nodeType === 'labjack-current') {
    if (!linkedChannel(node.id, 'channel', graph, incoming)) {
      issues.push(issue('error', node.id, 'Channel requires a channel reference'));
    }
    const shunt = linkedNode(node.id, 'shunt', graph, incoming);
    if (shunt && (shunt.nodeType !== 'constant' || shunt.config?.unit !== 'Ω')) {
      issues.push(issue('error', node.id, 'Shunt input must use Ω'));
    } else if (shunt && !(Number(shunt.config?.value) > 0)) {
      issues.push(issue('error', node.id, 'Shunt resistance must be positive'));
    } else if (!shunt && !(Number(node.config?.shuntOhms) > 0)) {
      issues.push(issue('error', node.id, 'Enter a positive shunt resistance or connect an Ω constant'));
    }
  }
  if (node.nodeType === 'labjack-ain') {
    const source = linkedNode(node.id, 'channel', graph, incoming);
    if (!['labjack-channel', 'labjack-channel-pair'].includes(source?.nodeType)) {
      issues.push(issue('error', node.id, 'Channel requires a channel reference or channel pair'));
    }
  }
  if (node.nodeType === 'labjack-thermocouple') {
    const source = linkedNode(node.id, 'pair', graph, incoming);
    if (source?.nodeType !== 'labjack-channel-pair') {
      issues.push(issue('error', node.id, 'Thermocouple requires a channel pair'));
    }
  }
  if (![10, 1, 0.1, 0.01].includes(Number(config.rangeV))) {
    issues.push(issue('error', node.id, 'Input range must be ±10, ±1, ±0.1, or ±0.01 V'));
  }
}

function linkedChannel(nodeId, pinId, graph, incoming) {
  const source = linkedNode(nodeId, pinId, graph, incoming);
  return source?.nodeType === 'labjack-channel' ? source.config?.channel ?? null : null;
}

function linkedNode(nodeId, pinId, graph, incoming) {
  const link = incoming.get(`${nodeId}:${pinId}`);
  if (!link) return null;
  return (graph.nodes ?? []).find((node) => node.id === link.fromNode) ?? null;
}

function linkedOrLiteral(node, pinId, graph, incoming, literal) {
  const source = linkedNode(node.id, pinId, graph, incoming);
  if (source?.nodeType === 'constant') return source.config?.value;
  return literal;
}

function issue(severity, subject, message) {
  return { severity, subject, message };
}

export function blockingIssues(issues) {
  return issues.filter((item) => item.severity === 'error');
}

function requiredNumber(value) {
  return value !== null && value !== '' && value !== undefined && Number.isFinite(Number(value));
}

File: base_station/src/base_station/web/daq_config/validation.py
"""Server-side validation for DAQ blueprint documents."""

from __future__ import annotations

import re
from math import isfinite

from base_station.web.daq_config.link_validation import validate_link_types
from base_station.web.daq_config.node_specs import validate_spec_node

AIN_PATTERN = re.compile(r"^AIN(\d{1,3})$")
MEASUREMENT_TYPES = {"labjack-ain", "labjack-current", "labjack-thermocouple"}
VALID_RANGES = {10.0, 1.0, 0.1, 0.01}
THERMOCOUPLE_TYPES = {"B", "E", "J", "K", "N", "R", "S", "T", "C"}


def validate_graph(graph: object) -> list[dict[str, str]]:
    issues: list[dict[str, str]] = []
    if not isinstance(graph, dict):
        return [_issue("graph", "Configuration must be a JSON object")]
    nodes = graph.get("nodes")
    links = graph.get("links")
    if not isinstance(nodes, list) or not isinstance(links, list):
        return [_issue("graph", "Configuration requires nodes[] and links[]")]
    if len(nodes) > 250 or len(links) > 500:
        issues.append(_issue("graph", "Configuration is larger than the supported editor limit"))
    metadata = graph.get("metadata") if isinstance(graph.get("metadata"), dict) else {}
    scan_rate = metadata.get("scanRate", 1000)
    if not isinstance(scan_rate, (int, float)) or not 1 <= scan_rate <= 100_000:
        issues.append(_issue("graph", "Scan rate must be between 1 and 100,000 samples/s"))
    resolution = metadata.get("streamResolutionIndex", 0)
    if not isinstance(resolution, (int, float)) or int(resolution) not in range(9):
        issues.append(_issue("graph", "Stream resolution must be Auto or index 1 through 8"))
    settling = metadata.get("streamSettlingUs", 0)
    if not isinstance(settling, (int, float)) or settling < 0:
        issues.append(_issue("graph", "Stream settling time cannot be negative"))

    node_map: dict[str, dict] = {}
    for node in nodes:
        if not isinstance(node, dict) or not isinstance(node.get("id"), str):
            issues.append(_issue("node", "Every node requires a string id"))
            continue
        node_id = node["id"]
        if node_id in node_map:
            issues.append(_issue(node_id, "Duplicate node id"))
        node_map[node_id] = node
        if not isinstance(node.get("pins"), list):
            issues.append(_issue(node_id, "Node pins must be an array"))

    for link in links:
        if not isinstance(link, dict):
            issues.append(_issue("link", "Every link must be an object"))
            continue
        _validate_link(link, node_map, issues)
    issues.extend(validate_link_types(graph))

    incoming = _incoming_links(links)
    mux_enabled = bool(metadata.get("mux80Enabled", False))
    for node in nodes:
        if not isinstance(node, dict):
            continue
        _validate_required_inputs(node, incoming, issues)
        if node.get("nodeType") in {"labjack-channel", "labjack-channel-pair"}:
            _validate_channel(node, mux_enabled, issues)
        if node.get("nodeType") in MEASUREMENT_TYPES:
            _validate_measurement(node, node_map, incoming, issues)
        spec_messages = validate_spec_node(node)
        if spec_messages is not None:
            issues.extend(_issue(node["id"], message) for message in spec_messages)
        _validate_transform(node, node_map, incoming, issues)
    return issues


def blocking_issues(issues: list[dict[str, str]]) -> list[dict[str, str]]:
    return [issue for issue in issues if issue.get("severity") == "error"]


def _validate_required_inputs(
    node: dict,
    incoming: dict[str, dict[str, dict]],
    issues: list[dict[str, str]],
) -> None:
    node_id = node.get("id")
    if not isinstance(node_id, str):
        return
    connected = incoming.get(node_id, {})
    for pin in node.get("pins", []):
        if not isinstance(pin, dict) or pin.get("direction") != "input" or pin.get("optional"):
            continue
        pin_id = pin.get("id")
        if pin_id not in connected:
            label = str(pin.get("label") or pin_id or "Input")
            issues.append(_issue(node_id, f"{label} is not connected", severity="warning"))


def _validate_link(link: dict, nodes: dict[str, dict], issues: list[dict[str, str]]) -> None:
    source = nodes.get(link.get("fromNode"))
    target = nodes.get(link.get("toNode"))
    if source is None or target is None:
        issues.append(_issue(str(link.get("id", "link")), "Link references a missing node"))
        return
    source_pins = {pin.get("id") for pin in source.get("pins", []) if isinstance(pin, dict)}
    target_pins = {pin.get("id") for pin in target.get("pins", []) if isinstance(pin, dict)}
    if link.get("fromPin") not in source_pins or link.get("toPin") not in target_pins:
        issues.append(_issue(str(link.get("id", "link")), "Link references a missing pin"))


def _validate_channel(node: dict, mux_enabled: bool, issues: list[dict[str, str]]) -> None:
    config = node.get("config") if isinstance(node.get("config"), dict) else {}
    channel = config.get("channel")
    number = _ain_number(channel)
    if number is None:
        issues.append(_issue(node["id"], "Select a valid LabJack AIN channel"))
        return
    valid = 0 <= number <= 13 or (mux_enabled and 48 <= number <= 127)
    if not valid:
        issues.append(_issue(node["id"], "AIN channel is unavailable for the selected hardware"))
    if mux_enabled and 4 <= number <= 13:
        issues.append(_issue(node["id"], "AIN4-AIN13 are occupied when MUX80 is enabled"))
    if node.get("nodeType") == "labjack-channel-pair" and not _valid_differential_positive(number):
        issues.append(_issue(node["id"], "Selected AIN cannot start a differential pair"))



def _validate_measurement(
    node: dict,
    nodes: dict[str, dict],
    incoming: dict[str, dict[str, dict]],
    issues: list[dict[str, str]],
) -> None:
    config = node.get("config") if isinstance(node.get("config"), dict) else {}
    if float(config.get("rangeV", 0.1)) not in VALID_RANGES:
        issues.append(_issue(node["id"], "AIN range must be one of ±10, ±1, ±0.1, or ±0.01 V"))
    node_type = node.get("nodeType")
    if node_type == "labjack-current":
        _require_channel_ref(node, "channel", nodes, incoming, issues)
        shunt = _linked_constant(node, "shunt", nodes, incoming)
        literal_shunt = config.get("shuntOhms")
        if shunt is not None and shunt.get("config", {}).get("unit") != "Ω":
            issues.append(_issue(node["id"], "Shunt input must use Ω"))
        elif shunt is not None and not _positive_number(shunt.get("config", {}).get("value")):
            issues.append(_issue(node["id"], "Shunt resistance must be positive"))
        elif shunt is None and not _positive_number(literal_shunt):
            issues.append(_issue(node["id"], "Enter a positive shunt resistance or connect an Ω constant"))
        return

    if node_type == "labjack-ain":
        source = _linked_node(node, "channel", nodes, incoming)
        if source is None or source.get("nodeType") not in {"labjack-channel", "labjack-channel-pair"}:
            issues.append(_issue(node["id"], "Channel requires a channel reference or channel pair"))
    if node_type == "labjack-thermocouple":
        source = _linked_node(node, "pair", nodes, incoming)
        if source is None or source.get("nodeType") != "labjack-channel-pair":
            issues.append(_issue(node["id"], "Thermocouple requires a channel pair"))
        if config.get("thermocoupleType") not in THERMOCOUPLE_TYPES:
            issues.append(_issue(node["id"], "Unsupported thermocouple type"))


def _validate_transform(
    node: dict,
    nodes: dict[str, dict],
    incoming: dict[str, dict[str, dict]],
    issues: list[dict[str, str]],
) -> None:
    config = node.get("config") if isinstance(node.get("config"), dict) else {}
    if node.get("nodeType") == "pressure-calibration":
        points = {
            key: _literal_or_constant(node, key, config.get(key), nodes, incoming)
            for key in ("inputMin", "inputMax", "psiMin", "psiMax")
        }
        if not all(_finite_number(value) for value in points.values()):
            issues.append(_issue(node["id"], "Pressure calibration requires all four calibration values"))
        elif points["inputMin"] == points["inputMax"]:
            issues.append(_issue(node["id"], "Pressure calibration input span cannot be zero"))
    if node.get("nodeType") == "load-cell":
        rated = _literal_or_constant(node, "ratedOutputMvV", config.get("ratedOutputMvV"), nodes, incoming)
        capacity = _literal_or_constant(node, "capacity", config.get("capacity"), nodes, incoming)
        zero = _literal_or_constant(node, "zeroV", config.get("zeroV"), nodes, incoming)
        excitation = _literal_or_constant(node, "excitation", config.get("excitationV"), nodes, incoming)
        if not _positive_number(rated):
            issues.append(_issue(node["id"], "Load cell rated output must be positive"))
        if not _positive_number(capacity):
            issues.append(_issue(node["id"], "Load cell rated capacity must be positive"))
        if not _finite_number(zero):
            issues.append(_issue(node["id"], "Load cell zero offset is required"))
        if not _positive_number(excitation):
            issues.append(_issue(node["id"], "Load cell excitation must be positive"))


def _incoming_links(links: list) -> dict[str, dict[str, dict]]:
    incoming: dict[str, dict[str, dict]] = {}
    for link in links:
        if isinstance(link, dict):
            incoming.setdefault(str(link.get("toNode")), {})[str(link.get("toPin"))] = link
    return incoming


def _linked_node(
    node: dict,
    pin: str,
    nodes: dict[str, dict],
    incoming: dict[str, dict[str, dict]],
) -> dict | None:
    link = incoming.get(node["id"], {}).get(pin)
    return nodes.get(link.get("fromNode")) if link else None


def _linked_channel(node: dict, pin: str, nodes: dict[str, dict], incoming: dict[str, dict[str, dict]]) -> dict | None:
    source = _linked_node(node, pin, nodes, incoming)
    return source if source and source.get("nodeType") == "labjack-channel" else None


def _linked_constant(node: dict, pin: str, nodes: dict[str, dict], incoming: dict[str, dict[str, dict]]) -> dict | None:
    source = _linked_node(node, pin, nodes, incoming)
    return source if source and source.get("nodeType") == "constant" else None


def _literal_or_constant(
    node: dict,
    pin: str,
    literal: object,
    nodes: dict[str, dict],
    incoming: dict[str, dict[str, dict]],
) -> object:
    constant = _linked_constant(node, pin, nodes, incoming)
    if constant is not None:
        return constant.get("config", {}).get("value")
    return literal


def _require_channel_ref(
    node: dict,
    pin: str,
    nodes: dict[str, dict],
    incoming: dict[str, dict[str, dict]],
    issues: list[dict[str, str]],
) -> dict | None:
    source = _linked_channel(node, pin, nodes, incoming)
    if source is None:
        issues.append(_issue(node["id"], f"{pin.title()} requires a channel reference"))
    return source


def _ain_number(value: object) -> int | None:
    if not isinstance(value, str):
        return None
    match = AIN_PATTERN.fullmatch(value)
    return int(match.group(1)) if match else None


def _differential_negative(positive: int) -> int:
    if positive >= 16:
        return positive + 8
    return positive + 1


def _valid_differential_positive(channel: int) -> bool:
    if channel < 16:
        return channel % 2 == 0
    if not 48 <= channel <= 127:
        return False
    return ((channel - 48) // 8) % 2 == 0


def _issue(subject: str, message: str, severity: str = "error") -> dict[str, str]:
    return {"severity": severity, "subject": subject, "message": message}


def _finite_number(value: object) -> bool:
    return isinstance(value, (int, float)) and isfinite(float(value))


def _positive_number(value: object) -> bool:
    return _finite_number(value) and float(value) > 0
